rotary_bed_MO.py

```diff
@@ -0,0 +1,252 @@
+jac_info.get_jac_list(
+    static_measurement_index,  # the index of SCMs in the jacobian array
+    dynamic_measurement_index,
+)  # the index of DCMs in the jacobian array
+
+# use MeasurementOptimizer to pre-compute the unit FIMs
+calculator = MeasurementOptimizer(
+    jac_info,  # SensitivityData object
+    measure_info,  # MeasurementData object
+    error_cov=error_mat,  # error covariance matrix
+    error_opt=CovarianceStructure.measure_correlation,  # error covariance options
+    print_level=3,  # I use highest here to see all information
+)
+
+# calculate a list of unit FIMs
+calculator.assemble_unit_fims()
+
+
+## MO optimization
+mip_option = True
+objective = ObjectiveLib.D
+fixed_nlp_opt = False
+mix_obj_option = False
+alpha_opt = 0.9
+
+sparse_opt = True
+fix_opt = False
+small_element = 0.0001  # the small element added to the diagonal of FIM
+file_store_name = "test_run_"  # # this is the file name string
+
+num_dynamic_time = np.linspace(2, 220, Nt)
+
+static_dynamic = [[9, 11], [10, 12]]  # These pairs cannot be chosen simultaneously
+time_interval_for_all = True
+
+# map the timepoint index to its real time
+dynamic_time_dict = {}
+for i, tim in enumerate(num_dynamic_time):
+    dynamic_time_dict[i] = tim
+
+# give range of budgets for this case
+budget_ranges = np.linspace(1000, 26000, 26)
+# give a trial ranges for a test; we use the first 3 budgets in budget_ranges
+trial_budget_ranges = budget_ranges[:3]
+# initialize with A-opt. MILP solutions
+# choose what solutions to initialize from:
+# minlp_D: initialize with minlp_D solutions
+# milp_A: initialize with milp_A solutions
+# lp_A: iniitalize with lp_A solution
+# nlp_D: initialize with nlp_D solution
+initializer_option = "milp_A"
+
+# budgets for the current results
+curr_results = np.linspace(1000, 26000, 26)
+
+# ==== initialization strategy ====
+# according to the initializer option, we provide different sets of initialization files
+if initializer_option == "milp_A":
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./rotary_results/Apr17_A_mip_", ""
+
+elif initializer_option == "minlp_D":
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./rotary_results/Dec7_", "_d_mip"
+
+elif initializer_option == "lp_A":
+    # initial solution file path and name
+    file_name_pre, file_name_end = "./rotary_results/May12_", "_a"
+
+elif initializer_option == "nlp_D":
+    file_name_pre, file_name_end = "./rotary_results/May10_", "_d"
+
+
+# initialize the initial solution dict. key: budget. value: initial solution file name
+# this initialization dictionary provides a more organized input format for initialization
+# Note: b can be float, such as 1000.0, but it is stored as 1000, so we int(b)
+initial_solution = {}
+# loop over budget
+for b in curr_results:
+    initial_solution[b] = file_name_pre + str(int(b)) + file_name_end
+
+
+# ===== run a test for a few budgets =====
```

Comment:
Instead, streamline this to remove duplicate code.

Reply:
done